### PR TITLE
CLI support for Documents Portal

### PR DESCRIPTION
**Pull request by polyatail**
This PR addresses the following issues, related to the new documents portal:

- [X] #209 Add CLI support for uploading, downloading, and listing files in the documents portal
- [X] #187 Enable export of PDF reports from notebooks to One Codex documents portal via `OneCodexDocumentExporter`
- Enable `OneCodexHTMLExporter` nbconvert entry point for exporting reports as HTML, in addition to PDF

### CODE REVIEW

File: onecodex/models/misc.py [hunk 1]
```diff
@@ -47,3 +52,76 @@ def search_public(cls, *filters, **keyword_filters):
...
+        Parameters
+        ----------
+        path : string, optional
+            Full path to save the file to. If omitted, defaults to the original filename
+            in the current working directory.
+        """
+        if path is None:
+            path = os.path.join(os.getcwd(), self.filename)
+
+        if os.path.exists(path):
```
Comment (audy) [inline onecodex/models/misc.py]: check if not `/dev/stdout`?

Comment (polyatail) [reply onecodex/models/misc.py]: Added

File: onecodex/lib/upload.py [hunk 2]
```diff
@@ -656,16 +660,179 @@ def upload_fileobj(file_obj, file_name, fields, session, samples_resource, log=N
...
+                semaphore.acquire()
+
+                try:
+                    file_uuid = upload_document_fileobj(*wrapped_args[:-1])
+                    uuids_completed.append(file_uuid)
+                except Exception as e:
+                    # handle inside the thread to prevent the exception message from leaking out
+                    wrapped_args[-1].value = '{}'.format(e)
+
+                    if log:
```
[the remark this answers is not in the document]
Comment (polyatail) [reply onecodex/lib/upload.py]: I don't think it's available when these methods (which are on the document models) are executed from, e.g., within a jupyter notebook.

File: onecodex/lib/upload.py [hunk 3]
```diff
@@ -656,16 +660,179 @@ def upload_fileobj(file_obj, file_name, fields, session, samples_resource, log=N
...
+    log : `logging.Logger`, optional
+        Used to write status messages to a file or terminal.
+    progressbar : `bool`, optional
+        If true, display a progress bar using Click.
+
+    Returns
+    -------
+    `list` of `string`s containing document UUIDs of newly uploaded files.
+    """
+    if threads is None:
```
[the remark this answers is not in the document]
Comment (polyatail) [reply onecodex/lib/upload.py]: Oooh this is leftover from when we were using DEFAULT_UPLOAD_THREADS.

File: onecodex/cli.py [hunk 4]
```diff
@@ -86,6 +88,116 @@ def scripts():
...
+@click.pass_context
+@pretty_errors
+@telemetry
+@login_required
+def documents_download(ctx, file, path):
+    doc_obj = ctx.obj['API'].Documents.get(file)
+
+    if not doc_obj:
+        log.error('Could not find document {} (404 status code)'.format(file))
+        return
```
[the remark this answers is not in the document]
Comment (polyatail) [reply onecodex/cli.py]: 👍

**Pull request merged**